Add IRBuilder::CreateFMA

frederik-h opened the pull request:
- Adjust some existing CreateIntrinsic uses appropriately
- Adjust IRBuilderTest.cpp and remove duplicate test case

frederik-h committed:
Add IRBuilder::CreateFMA

- Adjust some existing CreateIntrinsic uses appropriately
- Adjust IRBuilderTest.cpp and remove duplicate test case

arsenm commented on llvm/unittests/IR/IRBuilderTest.cpp:
```diff
@@ -109,21 +109,13 @@ TEST_F(IRBuilderTest, Intrinsics) {
-  EXPECT_FALSE(II->hasNoNaNs());
-
-  Result =
-      Builder.CreateIntrinsic(Intrinsic::fma, {V->getType()}, {V, V, V}, I);
```
Might as well test FMFSource while you're touching the test

frederik-h replied:
`I` is the `FMFSource`. Did you mean that I should more tests for it or did you mistake it for something else?

arsenm replied:
I more meant the new style where you directly supply the flags instead of a source instruction to copy from

frederik-h replied:
Got it now ;-).

arsenm commented on llvm/include/llvm/IR/IRBuilder.h:
```diff
@@ -1065,6 +1065,13 @@ class IRBuilderBase {
+  /// Create call to the fma intrinsic.
+  Value *CreateFMA(Value *Factor1, Value *Factor2, Value *Summand,
+                   FMFSource FMFSource = {}, const Twine &Name = "") {
+    return CreateIntrinsic(Intrinsic::fma, {Factor1->getType()},
```
Yes

frederik-h replied:
Ok, I have added it now.

frederik-h committed:
CreateFMA: Add assert to document that experimental.constrained.fma is not supported

frederik-h committed:
Add IRBuilder::CreateConstraintedFPIntrinsic

This is derived from CreateConstrainedFPBinOp but supports arbitrary
intrinsics.

frederik-h committed:
Add experimental_constrained_fma support to IRBuilder::CreateFMA

frederik-h committed:
IRBuilderTest: Pass FastMathFlags as FMFSource instead of using the Builder
